cpp/src/spatial/haversine.cu

```diff
@@ -0,0 +1,145 @@
+// /*
+//  * Copyright (c) 2020, NVIDIA CORPORATION.
+//  *
+//  * Licensed under the Apache License, Version 2.0 (the "License");
+//  * you may not use this file except in compliance with the License.
+//  * You may obtain a copy of the License at
+//  *
+//  *     http://www.apache.org/licenses/LICENSE-2.0
+//  *
+//  * Unless required by applicable law or agreed to in writing, software
+//  * distributed under the License is distributed on an "AS IS" BASIS,
+//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  * See the License for the specific language governing permissions and
+//  * limitations under the License.
+//  */
+
+#include <cuspatial/constants.hpp>
+#include <cuspatial/error.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+#include <cmath>
+#include <memory>
+#include <type_traits>
+
+namespace {
+
+template <typename T>
+__device__ T calculate_haversine_distance(T radius, T a_lon, T a_lat, T b_lon, T b_lat)
+{
+  auto ax = a_lon * DEGREE_TO_RADIAN;
+  auto ay = a_lat * DEGREE_TO_RADIAN;
+  auto bx = b_lon * DEGREE_TO_RADIAN;
+  auto by = b_lat * DEGREE_TO_RADIAN;
+
+  // haversine formula
+  auto x        = (bx - ax) / 2;
+  auto y        = (by - ay) / 2;
+  auto sinysqrd = std::sin(y) * std::sin(y);
+  auto sinxsqrd = std::sin(x) * std::sin(x);
+  auto scale    = std::cos(ay) * std::cos(by);
+
+  return 2 * radius * std::asin(std::sqrt(sinysqrd + sinxsqrd * scale));
+};
+
+struct haversine_functor {
+  template <typename T, typename... Args>
+  std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>> operator()(
+    Args&&...)
+  {
+    CUSPATIAL_FAIL("haversine_distance does not support non-floating-point types.");
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>> operator()(
+    cudf::column_view const& a_lon,
+    cudf::column_view const& a_lat,
+    cudf::column_view const& b_lon,
+    cudf::column_view const& b_lat,
+    double radius,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+  {
+    if (a_lon.is_empty()) { return cudf::experimental::empty_like(a_lon); }
+
+    auto mask_policy = cudf::experimental::mask_allocation_policy::NEVER;
+    auto result      = cudf::experimental::allocate_like(a_lon, a_lon.size(), mask_policy);
+
+    auto input_tuple = thrust::make_tuple(thrust::make_constant_iterator<T>(radius),
```

Comment:
does thrust have an overload that implicitly casts this from double to float when `T=float`? Or do we need this:

```suggestion
    auto input_tuple = thrust::make_tuple(thrust::make_constant_iterator<T>(static_cast<T>(radius)),
```